File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.cpp
class Solution {
public:
    string reverseWords(string s) {
        // Reverse the entire string
        reverse(s.begin(), s.end());

        int n = s.length();
        int i = 0;
        int l = 0;
        int r = 0;

        while (i < n) {
            // Skip leading spaces
            while (i < n && s[i] == ' ') {
                i++;
            }
            // Mark the start of the word
            l = r;
            // Copy the word
            while (i < n && s[i] != ' ') {
                s[r++] = s[i++];
            }
            // Reverse the word
            if (l < r) {
                reverse(s.begin() + l, s.begin() + r);
                s[r++] = ' ';
            }
        }

        // Remove the trailing space if there is one
        if (r > 0 && s[r - 1] == ' ') {
            r--;
        }

        return s.substr(0, r);
    }
};

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.java
class Solution {
    public static void reverse(char arr[], int i, int j) {
        while (i < j) {
            char temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
            i++;
            j--;
        }
    }

    public String reverseWords(String s) {
        char[] input = s.toCharArray();
        int n = input.length;

        
        reverse(input, 0, n - 1);

        
        char[] arr = new char[n + 1]; 
        int i = 0, r = 0, l = 0;

        while (i < n) {
        
            while (i < n && input[i] == ' ') i++;

        
            while (i < n && input[i] != ' ') {
                arr[r++] = input[i++];
            }

        
            if (l < r) {
                reverse(arr, l, r - 1);

                
                arr[r++] = ' ';
                l = r;
            }
        }

        
        // if (r > 0 && arr[r - 1] == ' ') r--;

        return new String(arr, 0, r-1);
    }
}
